src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/PkcsPalOpenSsl.Encrypt.cs

```diff
@@ -69,15 +70,144 @@ internal sealed partial class PkcsPalOpenSsl : PkcsPal
+            //     version CMSVersion,
+            //     originatorInfo[0] IMPLICIT OriginatorInfo OPTIONAL,
+            //     recipientInfos RecipientInfos,
+            //     encryptedContentInfo EncryptedContentInfo,
+            //     unprotectedAttrs[1] IMPLICIT UnprotectedAttributes OPTIONAL }
+
+            byte[][] contentType = originalMessage.ReadAndSplitNextEncodedValue();
+            DerSequenceReader envelopedDataReader = originalMessage.ReadExplicitContextSequence();
+
+            int versionNumber = envelopedDataReader.ReadInteger();
+
+            if (versionNumber > 4)
+                throw new PlatformNotSupportedException(SR.Cryptography_Cms_EncodeUnsupportedVersionPlatformNotSupported);
+
+            // "If unprotectedAttrs is present, then version shall be 2." - RFC 2630
+            if (versionNumber == 0)
+                versionNumber = 2;
+
+            byte[] version = ConcatenateArrays(DerEncoder.SegmentedEncodeUnsignedInteger(new byte[] { (byte)versionNumber }));
+
+            // We don't need to parse the OriginatorInfo, RecipientInfos, or EncryptedContentInfo, so
+            // we can just extract it to concatenate it.
+            byte[] remainingDataOfSequence = envelopedDataReader.ReadRemainingData();
+
+            byte[][] envelopedData = DerEncoder.ConstructSegmentedExplicitSequenceFromPayload(
+                0 /* Context number */,
+                version,
+                remainingDataOfSequence,
+                EncodeUnprotectedAttributes(unprotectedAttributes));
+
+            return DerEncoder.ConstructSequence(
+                contentType,
+                envelopedData);
+        }
+
+        private static byte[] ConcatenateArrays(byte[][] data)
+        {
+            int length = 0;
+            foreach (byte[] innerData in data)
+            {
+                length += innerData.Length;
+            }
+
+            byte[] concatenatedArray = new byte[length];
+
+            int offset = 0;
+            foreach(byte[] innerData in data)
+            {
+                Buffer.BlockCopy(innerData, 0, concatenatedArray, offset, innerData.Length);
+                offset += innerData.Length;
+            }
+
+            return concatenatedArray;
+        }
+
+        private byte[] EncodeUnprotectedAttributes(CryptographicAttributeObjectCollection unprotectedAttributes)
+        {
+            // In EnvelopedData UnprotectedAttributes are included as:
+            //
+            // unprotectedAttrs[1] IMPLICIT UnprotectedAttributes OPTIONAL
+            //
+            // UnprotectedAttributes ::= SET SIZE (1..MAX) OF Attribute
+
+            List<byte[]> setOfAttrs = new List<byte[]>(unprotectedAttributes.Count);
+
+            foreach (CryptographicAttributeObject attribute in unprotectedAttributes)
+            {
+                setOfAttrs.Add(EncodeAttribute(attribute));
+            }
+
+            setOfAttrs.Sort((a, b) => CompareByteArrays(a, b));
```

Comment:
Why is this a List<byte[]> rather than a byte[]?  It looks like we know ahead of time how many items there will be, in which case this is incurring several unnecessary allocations and copies.  It'd be better as:
```C#
var setOfAttrs = new byte[unprotectedAttributes.Count][];
for (int i = 0; i < setOfAttrs.Length; i++)
{
    setOfAttrs[i] = EncodeAttribute(unprotectedAttributes[i]);
}
```
and then just using ```setOfAttrs``` rather than ```setOfAttrs.ToArray()```.